src/react/useMutation.ts

```diff
@@ -0,0 +1,196 @@
+import React from 'react'
+
+import { useConfigContext } from './ReactQueryConfigProvider'
+import { useGetLatest, useMountedCallback } from './utils'
+import { Console, uid, noop } from '../core/utils'
+import {
+  QueryStatus,
+  MutationResultPair,
+  MutationFunction,
+  MutationConfig,
+} from '../core/types'
+
+// TYPES
+
+type Reducer<S, A> = (prevState: S, action: A) => S
+
+interface State<TResult, TError> {
+  status: QueryStatus
+  data?: TResult
+  error: TError | null
+}
+
+enum ActionType {
+  Reset = 'Reset',
+  Loading = 'Loading',
+  Resolve = 'Resolve',
+  Reject = 'Reject',
+}
+
+interface ResetAction {
+  type: ActionType.Reset
+}
+
+interface LoadingAction {
+  type: ActionType.Loading
+}
+
+interface ResolveAction<TResult> {
+  type: ActionType.Resolve
+  data: TResult
+}
+
+interface RejectAction<TError> {
+  type: ActionType.Reject
+  error: TError
+}
+
+type Action<TResult, TError> =
+  | ResetAction
+  | LoadingAction
+  | ResolveAction<TResult>
+  | RejectAction<TError>
+
+// HOOK
+
+const getDefaultState = (): State<any, any> => ({
+  status: QueryStatus.Idle,
+  data: undefined,
+  error: null,
+})
+
+function mutationReducer<TResult, TError>(
+  _state: State<TResult, TError>,
+  action: Action<TResult, TError>
+) {
+  switch (action.type) {
+    case ActionType.Reset:
+      return getDefaultState()
+    case ActionType.Loading:
+      return {
+        status: QueryStatus.Loading,
+      }
+    case ActionType.Resolve:
+      return {
+        status: QueryStatus.Success,
+        data: action.data,
+      }
+    case ActionType.Reject:
+      return {
+        status: QueryStatus.Error,
+        error: action.error,
+      }
+    default:
+      throw new Error()
```

Comment:
Is this case reachable? Does TS complain about not having it?

Reply:
The case should not be reachable but TS does not complain about it, changed to `return state` though to keep it in line with regular reducers